server/controller_client.go

```diff
@@ -0,0 +1,231 @@
+		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		return
+	}
+}
+
+// start an individual input stream
+func (c *clientHandler) startStream(unit *client.Unit) {
+	// when we have individual input streams, that'll go here.
+	c.log.Debugf("Got unit stream for processor: %s", unit.ID())
+	_ = unit.UpdateState(client.UnitStateHealthy, "healthy", nil)
+
+}
+
+// update an individual input stream
+func (c *clientHandler) updateStream(unit *client.Unit) {
+	// when we have individual input streams, that'll go here.
+	c.log.Debugf("Got unit input update for processor: %s", unit.ID())
+	_ = unit.UpdateState(client.UnitStateConfiguring, "updating", nil)
+	_ = unit.UpdateState(client.UnitStateHealthy, "healthy", nil)
+
+}
+
+// handle the UnitChangedAdded event from the V2 API
+func (c *clientHandler) handleUnitAdded(unit *client.Unit) {
+	c.addUnit(unit)
+	unitType := unit.Type()
+
+	if unitType == client.UnitTypeOutput { // unit startup for the shipper itself
+		c.setShipperUnitID(unit)
+		go c.startShipper(unit)
+	}
+	if unitType == client.UnitTypeInput { // unit startup for streams? Processors? Queue?
+		go c.startStream(unit)
+	}
+}
+
+// handle the UnitChangedModified event from the V2 API
+func (c *clientHandler) handleUnitUpdated(unit *client.Unit) {
+	c.log.Debugf("Got Unit Modified: %s", unit.ID())
+	unitType := unit.Type()
+
+	if unitType == client.UnitTypeOutput {
+		state, _ := unit.Expected()
+		if state == client.UnitStateStopped {
+			c.shutdown(unit)
+		}
+	} else {
+		c.updateStream(unit)
+	}
+}
+
+// a blocking call that will wait for the shipper components to gracefully shutdown, then send the unit update
+func (c *clientHandler) shutdown(shipperUnit *client.Unit) {
+	if c.shipperIsStopping {
+		return
+	}
+	_ = shipperUnit.UpdateState(client.UnitStateStopping, "shutting down shipper output", nil)
+	c.stopShipper()
+	// The server has successfully shut down
+	// In theory we want this to block, as we should wait for queues & outputs to empty.
+	c.shutdownComplete.Wait()
+	c.log.Debugf("Shutdown complete, sending STOPPPED")
+	_ = shipperUnit.UpdateState(client.UnitStateStopped, "Stopped shipper output", nil)
+}
+
+// runController is the main runloop for the shipper itself, and managed communication with the agent.
+func runController(ctx context.Context, agentClient client.V2) error {
+	log := logp.L()
+
+	err := agentClient.Start(ctx)
+	if err != nil {
+		return fmt.Errorf("error starting connection to client")
+	}
+
+	log.Debugf("Starting error reporter")
+	go reportErrors(ctx, agentClient)
+
+	log.Debugf("Started client, waiting")
+
+	handler := newClientHandler()
+
+	// receive the units
+	for {
+		select {
+		case <-ctx.Done():
+			handler.log.Debugf("Got context done")
+			shipperUnit := handler.getUnit(handler.shipperOutputID)
+			handler.shutdown(shipperUnit)
+			// If we get context done, just hard-stop
+			return nil
+
+		case change := <-agentClient.UnitChanges():
```

Comment:
Does this always unblock? Looks a bit dangerous because we don't have a timeout.

Reply:
Yah, for both of these cases, I'm kind of relying on that passed-in context to timeout the `select` statement if needed, since we need to read from the `UnitChanges` channel lest we block the client.

For context, most of this code is based off the client example in the `elastic-agent-client`: https://github.com/elastic/elastic-agent-client/blob/1a71765a8bbeeedb3e76d984c9f4f39bc9456e3c/pkg/client/client_v2_test.go#L264

Reply:
It looks like this is supposed to be an endless loop until the process exists, so it should be fine. First I thought we were to run this function repeatedly.